### PR TITLE
Config structure

### PR DESCRIPTION
**Pull request by PososikTeam**

### CODE REVIEW

File: src/constructor/constructor.py [hunk 1]
```diff
@@ -90,7 +91,14 @@ def __create_optimizer(parameters: Union[Module, Tensor, List[Union[Module, Tens
     def __create_scheduler(optimizer: Optimizer, scheduler_params: DictConfig) -> Dict[str, Any]:
         scheduler_class = SCHEDULERS.get(scheduler_params.name)
         scheduler = scheduler_class(optimizer, **scheduler_params.params)
-        pl_params = scheduler_params.pl_params if 'pl_params' in scheduler_params else {}
+        pl_params = scheduler_params.pl_params
+        # Omega conf error when create pl_params with only one argument, it's request second,
```
Comment (VladVin) [inline src/constructor/constructor.py:95]: Didn't get this comment, what is the argument?

File: src/constructor/config_structure.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+class LossParams:
+    name: str
+    tag: str
+    mapping: Dict[str, str]
+    params: Dict = field(default_factory=dict)
+    weight: Optional[float] = None
+
+@dataclass
+class JointLossParams:
+    loss_params: List[LossParams]
```
Comment (VladVin) [inline src/constructor/config_structure.py:82]: Let's rename it to `losses` and the outer name in config to `joint_loss`. It might be easier for a user to check if he needs to understand behaviour of the joint loss thing

File: src/metrics/metrics_manager.py [hunk 3]
```diff
@@ -94,23 +45,21 @@ def compute(self):
...
+        """Initialize MetricManager.
+
+        Args:
+            params: Metric parameters.
+        """
         super().__init__()
         # Change list to set in phases
         for param in params:
-            param.phases = set(param.phases)
+            param.phases = list(set(param.phases))
```
Comment (VladVin) [inline src/metrics/metrics_manager.py:58]: Do you still need set here?

File: tests/constructor/configs/config.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# defaults:
```
Comment (VladVin) [inline tests/constructor/configs/config.yaml:1]: What does this comment mean?

File: tests/constructor/test_config_structure.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+from typing import *
+
+from hydra.core.config_store import ConfigStore
+from omegaconf import OmegaConf
+
+from src.constructor.config_structure import ConfigParams, Phase
+
+
+@hydra.main(config_path="configs/", config_name="config.yaml")
+def test_config_load_when_full_config_was_define(config):
```
Comment (VladVin) [inline tests/constructor/test_config_structure.py:16]: was_defined

File: tests/constructor/test_config_structure.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    config_params = ConfigParams(**config)
+    # test if metric phase is enum
+    if isinstance(type(config_params.metrics[0].phases[0]), Phase):
+        print('Failed. Phase is not Enum.')
```
Comment (VladVin) [inline tests/constructor/test_config_structure.py:20]: This isn't a fail, this is just a warning. If you want the test to fail here, you should use an assert statement

File: tests/constructor/test_config_structure.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    else:
+        print('Phase enum OK.')
```
Comment (VladVin) [inline tests/constructor/test_config_structure.py:22]: I don't see any benefits of using prints in tests

File: tests/constructor/test_config_structure.py (hunk 5 above)
Comment (VladVin) [reply tests/constructor/test_config_structure.py:16]: Also, the name of the test suggests a big test, but in fact only two parameters are checked. Shouldn't you check all the parameters that you expect with a bunch of assert statements? (and check the comment below)

File: tests/constructor/test_config_structure.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    # test normalize_weights parameter
+    if config_params.losses.normalize_weights != False:
+        print('Failed. normalize_weights != False.')
+    else:
+        print('normalize_weights OK.')
+
+
+@hydra.main(config_path="configs/", config_name="config_without_metrics.yaml")
+def test_config_load_when_metric_is_not_define(config):
```
Comment (VladVin) [inline tests/constructor/test_config_structure.py:31]: Maybe you should just leave one giant test as `test_config_load_when_full_config_was_defined` or you can split this one by many small tests where every group of parameters is checked independently, so that it will be easier for contributors to find errors

File: tests/constructor/test_config_structure.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    if len(params) == 0:
+        print('Load empty params is OK.')
+    else:
+        print('Failed. Load empty params.')
+
+
+if __name__ == "__main__":
+    cs = ConfigStore.instance()
+    cs.store(name="config", node=ConfigParams)
+    test_config_load_when_full_config_was_define()
```
Comment (VladVin) [inline tests/constructor/test_config_structure.py:53]: Looks like these tests won't be captured by unittest

File: tests/constructor/test_constructor.py [hunk 10]
```diff
@@ -229,29 +232,32 @@ def __test_transforms_augmentations(self, kind):
...
-                    'params': {},
-                    'tag': 'classification',
-                    'mapping': {
-                        'input': 'logits',
-                        'target': 'target'
-                    },
-                    'weight': 0.7
-                }
-            ]
+            'losses':{
```
Comment (VladVin) [inline tests/constructor/test_constructor.py:235]: Did you run `flake8 <path/to/source/file>` ?

File: tests/constructor/test_constructor_with_conf_structure.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
+import sys
```
Comment (VladVin) [inline tests/constructor/test_constructor_with_conf_structure.py:1]: I see you pass the same structure with which `Constructor` was already tested in the `tests/constructor/test_constructor.py`. I think we don't need these tests, because config structure and constructor should be tested independently

File: train.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+from hydra.core.config_store import ConfigStore
+from pathlib import Path
+
+
+# Hack to fix multiprocessing deadlock when PyTorch's DataLoader is used
+# (more info: https://github.com/pytorch/pytorch/issues/1355)
+import cv2
+cv2.setNumThreads(0)
+
+# from src.constructor import TASKS
```
Comment (VladVin) [inline train.py:15]: Remove

File: train.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+from src.constructor.config_structure import ConfigParams
+# from src.constructor import create_trainer
+
+
+@hydra.main()
+def load_config(config):
+    # Need to add --config-path (-cp) and --config_name (-cn) in run command
```
Comment (VladVin) [inline train.py:22]: `config-name`

File: train.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    # Example config in configs/classification_cifar10.yaml
```
Comment (VladVin) [inline train.py:23]: Examples should be in `examples` folder

File: train.py [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    # Then the command line will be:
+    # python --config-path configs --config_name classification_cifar10 
+    config_params = ConfigParams(**config)
+    return config_params
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-cp', '--config-path', type=str,
```
Comment (VladVin) [inline train.py:31]: I think we should support just `config` parameter directing to the full config. It will be the only required argument

File: train.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    parser.add_argument('-cn', '--config-name', type=str,
+                        help="Yaml file name with configuration parameters.")
+    parser.add_argument('-jl', '--job-link', type=str,
+                        help="sagemaker job name, if running localy set to 'local'", default='local')
+    args = parser.parse_args()
+    
+    # getconfig name
+    config_name = args.config_name
+    config_name = config_name.split('.')[0] if '.' in config_name else config_name
+    structure_name = 'base_' + config_name
```
Comment (VladVin) [inline train.py:42]: Why do we need this?

File: train.py [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+@hydra.main()
+def load_config(config):
+    # Need to add --config-path (-cp) and --config_name (-cn) in run command
+    # Example config in configs/classification_cifar10.yaml
+    # Then the command line will be:
+    # python --config-path configs --config_name classification_cifar10 
+    config_params = ConfigParams(**config)
+    return config_params
+
+if __name__ == '__main__':
```
Comment (VladVin) [inline train.py:29]: Can the config parameters be overriden by cli arguments?

File: train.py (hunk 14 above)
Comment (VladVin) [inline train.py:23]: Put "TODO: add config into the project"

**Review by VladVin — CHANGES_REQUESTED** [review]

File: tests/constructor/configs/config.yaml (hunk 4 above)
Comment (PososikTeam) [reply tests/constructor/configs/config.yaml:1]: Same reason as ```'base_' + config_name```

https://hydra.cc/docs/upgrades/1.0_to_1.1/automatic_schema_matching/#:~:text=base_%20prefix%2C%20e.g.%20base_mysql.

File: train.py (hunk 17 above)
Comment (PososikTeam) [reply train.py:29]: Yes hydra support change any config parameters by use cli

File: train.py (hunk 15 above)
Comment (PososikTeam) [reply train.py:31]: Hydra not support only name it need path to

File: train.py (hunk 17 above)
Comment (PososikTeam) [reply train.py:31]: Actually no =). I need to remove argparser to give hydra ability change config by cli

Comment (VladVin) [reply train.py:31]: Ok, let's do it then

File: train.py (hunk 15 above)
Comment (VladVin) [reply train.py:34]: But you aren't using `hydra.main`. Ok, let's move to the correct usage of hydra

File: tests/constructor/test_hydra_load.py [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+import omegaconf
+from omegaconf import OmegaConf, DictConfig
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import *
+
+from src.constructor.config_structure import ConfigParams, Phase
+
+
+@hydra.main(config_path='/workdir/rbayazitov/torchok/examples/configs/', config_name='example')
```
Comment (VladVin) [inline tests/constructor/test_hydra_load.py:14]: Hard coded path.  
Overall, do you think we need this kind of test? Looks like you are testing Hydra's behaviour but not ours

File: tests/constructor/test_config_structure_load.py [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+def load_structured_config(path):
+    config = OmegaConf.load(path)
+    OmegaConf.resolve(config)
+    schema = OmegaConf.structured(ConfigParams)
+    config = OmegaConf.merge(schema, config)
+    return config
+
+
+class TestConfigSrucure(unittest.TestCase):
```
Comment (VladVin) [inline tests/constructor/test_config_structure_load.py:21]: `TestConfigStructure`

File: train.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+def main(config: DictConfig):
+    # Need to add --config-path (-cp) and --config_name (-cn) in run command
+    # Example config in configs/classification_cifar10.yaml
+    # TODO: add config into the project
+    # Then the command line will be:
+    # python --config-path configs --config_name classification_cifar10
+
+    # Resolve -> change evn variable to values for example ${oc.env:USER} -> 'root'
+    OmegaConf.resolve(config)
+    # Registrate structure
```
Comment (VladVin) [inline train.py:26]: Register

File: train.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+    schema = OmegaConf.structured(ConfigParams)
+    # Merge sturcture with config
+    config = OmegaConf.merge(schema, config)
+    # may be we don't need this line
```
Comment (VladVin) [inline train.py:30]: Need or don't need?

**Review by VladVin — CHANGES_REQUESTED** [review]
